### PR TITLE
Extended functionality for table widget

### PR DESCRIPTION
**Pull request by jo-mueller**
Hi Robert @haesleinhuepf ,

I made a few changes to the napari skimage regionprops table to fit it to my needs in the scope of [this](https://github.com/BiAPoL/napari-quantify-segmentation-quality) project. Here's a list of changes I made:

- Table widget now has a property `_name` by which a whidget can be retrieved when `get_table(...)` is called.
- `get_table(...)` now accepts both strings or napari layers as input:
```Python
def get_table(labels_layer: Union[napari.layers.Layer, str],
              viewer: napari.Viewer) -> TableWidget:
```
   This allows for table widgets to be re-used.

- Added more docstrings to the functions in case it should be desirable to document the API in the future.
- Added a function `append_table()` which allows to add data in the form of a dictionary to a given table widget instance.

These changes allow the table to be used in two ways: Tied to a labels layer so that the usage of napari skimage regionprops is fully conserved but also as an extendable table so that results from multiple images can be added.

In my usecase, I use the table to compare two images and would like to repeat this several times but also compare different computations based on the same image. I thus added the names of the input images as columns in the table so that is always clear where the data comes from.

Let me know what you think :)

### CODE REVIEW

File: napari_skimage_regionprops/_table.py [hunk 1]
```diff
@@ -89,43 +103,104 @@ def set_content(self, table : dict):
...
+
     """
+    _table = table_widget._table
+
+    for key in _table.keys():
+        if key in table.keys():
+
+            _table[key] += list(table[key])
+        else:
+            _table[key] += list('NaN')
```
Comment (haesleinhuepf) [inline napari_skimage_regionprops/_table.py:165]: I think if one merges a table with columns A,B, C and a table with B, C, D, either A or D is lost. I suggest using pandas for merging tables. 
Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.append.html
Example: https://github.com/haesleinhuepf/scripts/blob/master/append_tables.ipynb

Comment (jo-mueller) [reply napari_skimage_regionprops/_table.py:165]: Yes, I think the current implementation will only keep columns that intersect with the ones present in the table. I was also thinking of sing pandas as table backend - should I add this to this PR or rather as a separate one?

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hi Johannes,

thanks for your suggestion! I like the `append` idea in particular. I'd like to ask for some changes though. You are suggesting three changes and I think they should be discussed separately. Hence, I recommend opening three new smaller PRs.

1. I think `append_table` would be better suited as a member function of the `Table` class. You can then call `get_table(layer).append(dictionary)`. Again, as suggested above, please use pandas. In the future, when the napari developers finished their work on the tables, we can directly call `layer.properties.append(other_pandas_table)`.
2. `get_table(str)` is a difficult API change which comes with some risks. Assume plugin A creates a table associated with a layer and plugin B grabs this table just because the name by chance fits. It might then manipulate the wrong table. Furthermore, if I understand your use-case right, I suggest to always store comparisons with a specific 'ground truth' layer in this particular layer. We can then also program another plugin, that consumes tables from multiple ground truth layers and analyses/compares them.
3. You are submitting many changes that look to me like format changes. I'm not sure if there are changes among them which change functionality. If you think these format changes are necessary, please submit them as independent pull-request. I assume you did those automatically, so if you think these changes are crucial, please add to the documentation how other contributors can make code fit to your suggested scheme.

Let me know if I can help.

Best,
Robert

**Comment by haesleinhuepf** [comment]
Also the tests are failing after your changes. Could you please check? Use `pytest --cov=napari_skimage_regionprops` from the root directory of the repository to run them and also see coverage changes.

**Comment by jo-mueller** [comment]
>I think append_table would be better suited as a member function of the Table class. You can then call get_table(layer).append(dictionary). Again, as suggested above, please use pandas. In the future, when the napari developers finished their work on the tables, we can directly call layer.properties.append(other_pandas_table).

Implemented in #13

**Comment by jo-mueller** [comment]
> get_table(str) is a difficult API change which comes with some risks. Assume plugin A creates a table associated with a layer and plugin B grabs this table just because the name by chance fits. It might then manipulate the wrong table. Furthermore, if I understand your use-case right, I suggest to always store comparisons with a specific 'ground truth' layer in this particular layer. We can then also program another plugin, that consumes tables from multiple ground truth layers and analyses/compares them.

On this one: Currently, if I execute napari-skimage regionprops twice (same input data), I already create this kind of ambiguity, don't I? 

`get_table()` will look for a table_widget associated with a certain layer - what happens if multiple widgets are associated with this layer? In this case it would be better if the plugin would check if an associated plugin already exists and to append/replace the data?

**Comment by haesleinhuepf** [comment]
> On this one: Currently, if I execute napari-skimage regionprops twice (same input data), I already create this kind of ambiguity, don't I?

If you call this ([source](https://github.com/haesleinhuepf/napari-simpleitk-image-processing/blob/main/src/napari_simpleitk_image_processing/_simpleitk_image_processing.py#L877-L881)):
```
labels_layer.properties = results
from napari_skimage_regionprops import add_table
add_table(labels_layer, napari_viewer)
```
... you know that afterwards, `labels_layer.properties` contains your data. When calling this code, you also know that you potentially overwrite something. You could read `layer.properties` in advance and ask the user, if it should be overwritten, appended, etc. That's just not implemented yet.

If you call
```
get_table('Whatever', napari_viewer).set_content(results)
```
You have no idea which layer's data was just overwritten.

The [idea of the table widgets](https://github.com/haesleinhuepf/napari-skimage-regionprops/blob/master/napari_skimage_regionprops/_table.py#L10) is really just a viewer for `layer.properties` data. Any kind of data manipulation should be done on `layer.properties` and it is great that the napari developers _force_ us to follow this standard, because in that way, all plugins are compatible with each other.

> would be better if the plugin would check if an associated plugin already exists and to append/replace the data?

I _think_ plugins should not check if other plugins exist. Plugins should be as independent as possible. It is true that one plugin can overwrite the results of another. But I would say that's not our problem for now and this should be fixed on a higher lever (e.g. in `layer.properties`, feel free to contribute to [the discussion here](https://github.com/napari/napari/issues/3708)), or it is the responsibility of the individual plugin developers. Is this a problem for your desired functionality? I'm not sure if I can see the issue.

By the way, at the moment all plugins just overwrite `layer.properties` per default, also because there is no other convenient option. That's why your idea of `append_table` is really great! We should have this.

**Comment by jo-mueller** [comment]
>I think plugins should not check if other plugins exist.

My bad, I meant to refer to tables: 
>[...] would be better if a plugin would check if an associated table already exists and to append/replace the data?

which sounds very similar to what you proposed further up:
>We can then also program another plugin, that consumes tables from multiple ground truth layers and analyses/compares them.

Currently, to do this I could
- create an empty dummy layer that hosts all the collected data
- Take any of the previously generated tables and append - which is a bit arbitrary because it is no longer obvious where the data lives.

It would probably be a nice solution to integrate this with the [napari project viewer](https://github.com/guiwitz/napari-annotation-project), but I'll have to scout this one a bit first.

**Comment by haesleinhuepf** [comment]
> create an empty dummy layer that hosts all the collected data

I would like to better understand for what. What's the use-case here?

**Comment by jo-mueller** [comment]
> I would like to better understand for what. What's the use-case here?

This is taken from the direction of the to-be-developed segmentation quantification plugin. It compares two labels images (ground truth/segmented image) and calculates a number of metrics for comparison. Ultimately, I would like for the plugin to allow repeating the analysis for an arbitrary number of label image pairs. I would then like to collect all acquired metrics in a single table for some really comprehensive analysis.

On that note: If I used a plugin to calculate some result which is stored in `layer.properties`. What would happen if I then used `napari-skimage-regionprops` on the same layer? Would the current properties be overwritten?

**Comment by haesleinhuepf** [comment]
> This is taken from the direction of the to-be-developed segmentation quantification plugin. It compares two labels images (ground truth/segmented image) and calculates a number of metrics for comparison. Ultimately, I would like for the plugin to allow repeating the analysis for an arbitrary number of label image pairs. I would then like to collect all acquired metrics in a single table for some really comprehensive analysis.

Let's talk about that in a meeting. I think it's possible :-)

**Comment by jo-mueller** [comment]
Would love to. Tomorrow (e.g., thursday?)

**Pull request closed without merging**